src/QirRuntime/public/ResourceStatistics.hpp

```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <sstream>
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    static const char* names[] = {"cX", "cCX", "cY",  "cCY",  "cZ", "cCZ", "cH",  "cCH",  "cR", "cCR",
+                                  "cS", "cCS", "cAS", "cCAS", "cT", "cCT", "cAT", "cCAT", "cM"};
+    static constexpr int counters = sizeof(names) / sizeof(char*);
+
+    struct ResourceStatistics
+    {
+        // these members must match names[] above
+        long cX, cCX = 0;

```

Comment:
Given that there's no methods to increment or read these counters, and that the counters are all public, I assume that means that users of this type just access and update the counters directly. This is very fragile and very difficult to extend and maintain.

Reply:
Yes, the direct access was intended here to allow LLVM to optimize the code away and collapse the loops to a simple increment instructions based on the loop count. I played with it at the beginning of the project when the focus was on incubation of highly optimized native resource tracer. I'll move it into experimental together with the TracingContext.